internal/xds/translator/extauth.go

```diff
@@ -0,0 +1,339 @@
+		}
+	}
+
+	if extAuth.HTTP != nil {
+		config.Services = &extauthv3.ExtAuthz_HttpService{
+			HttpService: httpService(extAuth.HTTP),
+		}
+	} else if extAuth.GRPC != nil {
+		config.Services = &extauthv3.ExtAuthz_GrpcService{
+			GrpcService: &corev3.GrpcService{
+				TargetSpecifier: &corev3.GrpcService_EnvoyGrpc_{
+					EnvoyGrpc: grpcService(extAuth.GRPC),
+				},
+				Timeout: &duration.Duration{
+					Seconds: defaultExtServiceRequestTimeout,
+				},
+			},
+		}
+	}
+
+	return config
+}
+
+func httpService(http *ir.HTTPExtAuthService) *extauthv3.HttpService {
+	var (
+		enableTLS        = http.TLS != nil
+		uri              string
+		headersToBackend []*matcherv3.StringMatcher
+	)
+
+	if enableTLS {
+		uri = fmt.Sprintf("https://%s:%d", http.Host, http.Port)
+	} else {
+		uri = fmt.Sprintf("http://%s:%d", http.Host, http.Port)
+	}
+	if http.Path != nil {
+		uri = fmt.Sprintf("%s%s", uri, *http.Path)
+	}
+
+	for _, header := range http.HeadersToBackend {
+		headersToBackend = append(headersToBackend, &matcherv3.StringMatcher{
+			MatchPattern: &matcherv3.StringMatcher_Exact{
+				Exact: header,
+			},
+		})
+	}
+
+	return &extauthv3.HttpService{
+		ServerUri: &corev3.HttpUri{
+			Uri: uri,
+			HttpUpstreamType: &corev3.HttpUri_Cluster{
+				Cluster: clusterName(http.Host, http.Port),
+			},
+			Timeout: &duration.Duration{
+				Seconds: defaultExtServiceRequestTimeout,
+			},
+		},
+		AuthorizationResponse: &extauthv3.AuthorizationResponse{
+			AllowedUpstreamHeaders: &matcherv3.ListStringMatcher{
+				Patterns: headersToBackend,
+			},
+		},
+	}
+}
+
+func grpcService(grpc *ir.GRPCExtAuthService) *corev3.GrpcService_EnvoyGrpc {
+	return &corev3.GrpcService_EnvoyGrpc{
+		ClusterName: clusterName(grpc.Host, grpc.Port),
+		Authority:   grpc.Host,
+	}
+}
+
+// routeContainsExtAuth returns true if ExtAuth exists for the provided route.
+func routeContainsExtAuth(irRoute *ir.HTTPRoute) bool {
+	if irRoute == nil {
+		return false
+	}
+
+	if irRoute != nil &&
+		irRoute.ExtAuth != nil {
+		return true
+	}
+
+	return false
+}
+
+func (*extAuth) patchResources(tCtx *types.ResourceVersionTable,
+	routes []*ir.HTTPRoute) error {
+	if tCtx == nil || tCtx.XdsResources == nil {
+		return errors.New("xds resource table is nil")
+	}
+
+	var errs error
+	for _, route := range routes {
+		if !routeContainsExtAuth(route) {
+			continue
+		}
+		if route.ExtAuth.HTTP != nil {
+			if err := createExtAuthCluster(
+				tCtx,
+				route.ExtAuth.HTTP.Host,
+				route.ExtAuth.HTTP.Port,
+				route.ExtAuth.HTTP.TLS); err != nil {
+				errs = errors.Join(errs, err)
+				continue
+			}
+		} else {
+			if err := createExtAuthCluster(
+				tCtx,
+				route.ExtAuth.GRPC.Host,
+				route.ExtAuth.GRPC.Port,
+				route.ExtAuth.GRPC.TLS); err != nil {
+				errs = errors.Join(errs, err)
+				continue
+			}
+		}
+	}
+
+	return errs
+}
+
+// createExtAuthCluster creates the cluster for the external auth service.
+func createExtAuthCluster(tCtx *types.ResourceVersionTable,
```

Comment:
I assume that multiple security policies may reference the same auth-server `host:port` combination. It seems that a cluster is identified by name.  Does this mean that `ErrXdsClusterExists` can be triggered very often? If multiple policies define different cluster settings (e.g. transport socket setting) for the same `host:port` auth server, only the first policy's settings will apply in practice?

Reply:
@guydc Good catch!  I believe different policies shouldn't have different TLS settings for the same host+port combination?

Reply:
@guydc Good catch!  I believe different policies shouldn't have different TLS settings for the same host+port combination?

Reply:
Yes, I totally agree that cluster deduplication is a good idea. And you're right that for TLS, I would expect that the same target port would be either tls-enabled or not. 

Still, consider the following situation: a TLS-enabled ext-auth server is referenced by policies A and B. Policy B is misconfigured (by mistake) by the end user to not use TLS. If Policy B is the first to be translated, it would also break auth-server integration for all routes that policy A is attached to (even though policy A is ok). 

In some cases, there could even be a legitimate reason to have different cluster settings for the same host:port. For example, if a specific host:port supports both HTTP and GRPC auth interfaces, policies may need to reference the same host:port with different cluster HTTP settings (e.g. enabling HTTP2 for grpc).